cmd/kata-local-ci

```diff
@@ -0,0 +1,260 @@
+    )
+
+    case "$distro" in
+	ubuntu)
+	    distro_name=ubuntu-18.04
+	    distro_commands=("${ubuntu_distro_commands[@]}")
+	    osvariant=ubuntu18.04
+	    ;;
+	centos)
+	    distro_name=centos-8.2
+	    osvariant=centos8
+	    distro_commands=("${rh_distro_commands[@]}")
+	    ;;
+	fedora)
+	    distro_name=fedora-32
+	    osvariant=fedora32
+	    distro_commands=("${rh_distro_commands[@]}")
+	    final_firstboot_cmd=("${disable_cgroupv2[@]}")
+	    ;;
+	*)
+	    echo "Distro $1 not recognized. Please choose ubuntu, centos or fedora"
+            exit 3
+    esac
+
+    if [  ! -z "$network" ]; then
+	network="--network $network"
+    fi
+
+    VM_NAME=kata-$distro-ci
+    progress "Removing earlier VM $VM_NAME"
+    echo "The script may now ask you for a password to setup the VM named $VM_NAME"
+    echo "You can avoid these prompts by executing:"
+    echo "sudo bash -c \"echo > /etc/sudoers.d/kata-local-ci '"$(whoami) "ALL=(ALL) NOPASSWD: /usr/bin/virsh,/usr/bin/virt-install,/usr/bin/virt-builder'\""
+
+    sys virsh destroy $VM_NAME || true
+    sys virsh undefine $VM_NAME --remove-all-storage  --snapshots-metadata || true
+
+    set -e
+    progress "Launching virt-builder"
+    sys virt-builder $distro_name \
+         -o $disk --format qcow2 --size=$size \
+	 --root-password password:kata \
+	 "${distro_commands[@]}" \
+	 --mkdir /home/kata \
+	 --delete /.autorelabel \
+	 --ssh-inject "root:file:$key" \
+	 --install "git,rsync,curl,make,tar,tc" \
+         "${firstboot_cmds[@]}" \
+         "${final_firstboot_cmd[@]}" \
+	 --password  'kata:password:kata' \
+	 --hostname kata-$distro-ci
+    progress "Launching virt-install"
+    sys virt-install --import --name $VM_NAME  --memory $memory --vcpus $cpus \
+         --disk path=$disk,format=qcow2 \
+         --os-variant=$osvariant --noautoconsole $network
+
+    echo -n "Waiting for guest to get an IP address"
+    while ! sudo virsh domifaddr $VM_NAME | grep -q vnet ; do echo -n .; sleep 1; done
+
+    VM_IPADDR=$(sudo virsh domifaddr $VM_NAME | grep vnet | sed -e 's|^\s*vnet[0-9]*.*ipv4\s*\([0-9.]*\)/[0-9]*|\1|g')
+    echo "Got $VM_IPADDR"
+    progress "Guest IP address is $VM_IPADDR"
+
+    progress "Waiting for ssh to become available"
+    while ! ssh -o StrictHostKeyChecking=no -o UpdateHostKeys=no kata@$VM_IPADDR "mkdir -p ~/go/src/github.com/kata-containers"; do
+        sleep 1
+    done
+
+    progress "Initial rsync of kata-containers directory"
+    run rsync -av $directory/kata-containers kata@$VM_IPADDR:$vmdir/
+
+    progress "Initial rsync of tests directory"
+    run rsync -av $directory/tests kata@$VM_IPADDR:$vmdir/
+
+    progress "Launching CI setup in $VM_NAME (this will take a while)"
+    ci_cmd "cd $vmdir/tests; export GOPATH=~/go; .ci/setup.sh"
+
+    cat > "$kci"/$distro-config <<EOF
+VM_NAME="$VM_NAME"
+VM_IPADDR=$VM_IPADDR
+VM_DISK=$disk
+EOF
+
+    progress "Taking snapshot of $VM_NAME"
+    ci_root_cmd "sync; sync; sync; sleep 3"
```

Comment:
Why are all these sleeps eneded?

Reply:
Normally, when you want to have a disk-only snapshot, you use an in-guest agent which ensures that the disk is in a consistent state. I had trouble getting a portable / consistent way to get the agent installed (at least without another VM reboot). So I ended up trying without, meaning full snapshot.

I observed that the resume was quite slow, and some quick experiments persuaded me that the system began writing stuff to disk on resume. So I added the line above in an attempt to quiesce disk writes _before_ taking the snapshot, so that we don't redo them each time we resume. Not sure if it helps at all.